File: dsa/li09_linkedlist/L19_RemoveNthNodeFromEndOfList.java
package li09_linkedlist;

import com.algos.topics.models.ListNode;

public class L19_RemoveNthNodeFromEndOfList {

	public static void main(String[] args) {
		L19_RemoveNthNodeFromEndOfList obj = new L19_RemoveNthNodeFromEndOfList();
		SingleLinkedList list = new SingleLinkedList();
		list.insertAfterEnd(1);
		list.insertAfterEnd(2);
		list.insertAfterEnd(3);
		list.insertAfterEnd(4);

		list.head = obj.removeNthFromEnd(list.head, 4);     // 2 3 4
		list.print();
	}

    // Problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
    // Idea: First move right ptr to n nodes. Then move both left and right ptr till right.right is null.
    // Now the left node which was pointing head node, points to n+1th node fron last. Remove nth node.
	public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode left = head, right = head;
        for (int cnt = 0; cnt < n; cnt++)
            right = right.next;

        if(right == null)       // Remove head case
            return head.next;

        right = right.next;     // As we want left ptr to be delayed by one node
        while(right != null){
            left = left.next;
            right = right.next;
        }

        left.next = left.next.next;
        return head;
    }

}

File: dsa/concepts/09_linkedlist/python/l19_remove_nth_node_from_end_of_list.py
"""
Problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
Idea:
First move right pointer by n nodes.
Then move both left and right pointers until right.next is null.
Now the left node points to (n+1)th node from the end.
Remove the nth node from the end.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def remove_nth_from_end(head: ListNode, n: int) -> ListNode:
    left = head
    right = head

    # Move right pointer n steps ahead
    for _ in range(n):
        right = right.next

    # If right is null, remove head
    if right is None:
        return head.next

    # Delay left pointer by one node
    right = right.next

    while right:
        left = left.next
        right = right.next

    # Remove nth node from end
    left.next = left.next.next
    return head


# Example usage
if __name__ == "__main__":
    # List: 1 -> 2 -> 3 -> 4
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))

    head = remove_nth_from_end(head, 4)

    # Print result: 2 -> 3 -> 4
    while head:
        print(head.val, end=" ")
        head = head.next
